src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonAfterTypeMemberDeclarationCheck.java

```diff
@@ -0,0 +1,178 @@
+            TokenTypes.INSTANCE_INIT,
+            TokenTypes.CTOR_DEF,
+            TokenTypes.METHOD_DEF,
+            TokenTypes.ENUM_CONSTANT_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        switch (ast.getType()) {
+            case TokenTypes.CLASS_DEF:
+            case TokenTypes.INTERFACE_DEF:
+            case TokenTypes.ENUM_DEF:
+            case TokenTypes.ANNOTATION_DEF:
+                checkTypeDefinition(ast);
+                break;
+            case TokenTypes.VARIABLE_DEF:
+                checkVariableDefinition(ast);
+                break;
+            case TokenTypes.ENUM_CONSTANT_DEF:
+                checkEnumConstant(ast);
+                break;
+            default:
+                checkTypeMember(ast);
+                break;
+        }
+    }
+
+    /**
+     * Checks if type member has unnecessary semicolon.
+     *
+     * @param ast type member
+     */
+    private void checkTypeMember(DetailAST ast) {
+        if (isSemicolon(ast.getNextSibling())) {
+            log(ast.getNextSibling(), MSG_SEMI);
+        }
+    }
+
+    /**
+     * Checks if type definition has unnecessary semicolon.
+     *
+     * @param ast type definition
+     */
+    private void checkTypeDefinition(DetailAST ast) {
+        if (!ScopeUtil.isOuterMostType(ast) && isSemicolon(ast.getNextSibling())) {
+            log(ast.getNextSibling(), MSG_SEMI);
+        }
+        final DetailAST firstMember =
+            ast.findFirstToken(TokenTypes.OBJBLOCK).getFirstChild().getNextSibling();
+        if (isSemicolon(firstMember) && !ScopeUtil.isInEnumBlock(firstMember)) {
+            log(firstMember, MSG_SEMI);
+        }
+    }
+
+    /**
+     * Checks if variable definition has unnecessary semicolon.
+     *
+     * @param ast variable definition
+     */
+    private void checkVariableDefinition(DetailAST ast) {
+        if (isFieldDeclaration(ast) && isSemicolon(ast.getNextSibling())) {
+            log(ast.getNextSibling(), MSG_SEMI);
+        }
+    }
+
+    /**
+     * Checks if enum constant has unnecessary semicolon.
+     *
+     * @param ast enum constant
+     */
+    private void checkEnumConstant(DetailAST ast) {
+        final DetailAST next = ast.getNextSibling();
+        if (isSemicolon(next) && isSemicolon(next.getNextSibling())) {
+            log(next.getNextSibling(), MSG_SEMI);
+        }
+    }
+
+    /**
+     * Checks that {@code ast} is a semicolon.
+     *
+     * @param ast token to check
+     * @return true if ast is semicolon, false otherwise
+     */
+    private static boolean isSemicolon(DetailAST ast) {
+        return ast.getType() == TokenTypes.SEMI;
+    }
+
+    /**
+     * Checks that {@code variableDef} is a field declaration.
+     *
+     * @param variableDef token to check
+     * @return true if variableDef is a field declaration, false otherwise
+     */
+    private static boolean isFieldDeclaration(DetailAST variableDef) {
```

Comment:
Method is named `isFieldDeclaration` but code just returns `isSemicolon`. Please find a better suitable method name, it doesn't check if the AST is field or not.

Reply:
It accepts variable definition token, as javadoc says.
Difference between field definition and local variable definition is a last token - for field it is semicolon, for local variable - something else.

So we check last child to determine is it field definition or not.

I do not see problem.

Reply:
The problem is the method is named for something where the statements of that method have nothing to do with that name. By having a name like `isFieldDeclaration` someone in the future could call this method assuming it is checking if the token given is a field and not realize it is doing something else without having to look at the code.
The sign of a good name is you should be able to look at the method name and javadoc and get an idea of what it is doing without having to look at the code. Both of those should explain the method, your method name doesn't do this because the method is not checking if something is field or not.

Reply:
I inlined this method